### PR TITLE
DB Replica for local development

### LINKED ISSUE #8: Sharing DB for development
@icyflame @amrav @DefCon-007 @athityakumar @thealphadollar

What is the best practice for providing read-access to the DB with anyone interesting in contributing to the project?
_aka_ How do I share a replica of the production DB for local development?

From the top of my mind:
- Creating a credential user for every individual with read access.
    - Cons:
        - Proxy connection under Institute's LAN is blocked.
        - Tedious process on the part of the project maintainer.
- Storing a database dump in the repository.
    - Cons:
        - Regularly update the dump.
        - Bloats the size of the repo.

### PR DESCRIPTION
**Pull request by kshitij10496**
Here is my 2-fold proposal for sharing DB for local testing and development:

### Exporting Production DB to a Dump
- DB Admin/Maintainer creates a backup of the DB from Heroku Dashboard.
- Update the dump in the repo using the naming format `hercules_backup_YYYYMMDD.dump`.
- The DB is available for everyone to import and dig aka directly open-sourced.

### Importing Dump Locally
- The prospective contributor creates a PostgreSQL database: 
`$ createdb hercules_testdb`
- Import the dump using `pg_restore`: 
`$ pg_restore --verbose --clean --no-acl --no-owner -h locahost -u $USER -d hercules_testdb path/to/dump`

Considerations:
- More read operations than write operations. Moreover, the write operations (inserting and updating) are comparatively infrequent given the fact that the database is mostly updated once a semester.

This is the simplest and the minimalist solution to the problem I could come up with.
I have unapologetically prioritised ease of setup and involvement from beginner developers (KGP community centric) over other factors such as maintainance, contraptions of 3rd party cloud services.

Basically, I considered the points raised by @icyflame in the discussions and incorporated my personal trade-off choices while coming up with the solution. 
References:
1. [Importing and Exporting Heroku Postgres Databases](https://devcenter.heroku.com/articles/heroku-postgres-import-export)
2. [How To Backup PostgreSQL Databases on an Ubuntu VPS](https://www.digitalocean.com/community/tutorials/how-to-backup-postgresql-databases-on-an-ubuntu-vps)

Resolves #8 

cc @icyflame @thealphadollar @themousepotato @Ayushk4

### CODE REVIEW

**Comment by icyflame** [comment]
@kshitij10496 Looks good. This is a good way to make it easier for developers to contribute to hercules!

**Comment by kshitij10496** [comment]
@icyflame Thanks for the support! ❤️ 

Being a pragmatist, I would request you guys to set up the database locally using the above method and share your feedback. This should help us evaluate the effectiveness of the proposed solution in practice. I plan on wrapping this up over the weekend and writing the Development Workflow. 😄

**Comment by thealphadollar** [comment]
LgTM too, :+1:  I think this should be merged ASAP. I would also suggest adding the steps explained in the PR to the readme file.

**Comment by kshitij10496** [comment]
Thanks @thealphadollar for the nod!

>  I would also suggest adding the steps explained in the PR to the readme file.

Sure. I will add this to the [Development Workflow](https://github.com/kshitij10496/hercules/wiki/Development-Workflow) wiki and link it in the README. 😃 

To rephrase what I said above, I would be more comfortable to merge this is if somebody can successfully import the DB using the steps I've outlined and report back here.

**Comment by kshitij10496** [comment]
I will be merging this in as soon as I get 2 review approvals or 24 hours pass, whichever happens, the earliest. 😄

**Review by thealphadollar — COMMENTED** [review]
I tried restoring using the provided commands, and surely setting it up isn't that easy. 

Faced the issue while using the `create_db` command and the following solution worked.
https://stackoverflow.com/questions/16973018/createuser-could-not-connect-to-database-postgres-fatal-role-tom-does-not-e

Later on, I was asked for password and I had created none. The solution was to replace `$USER` with `postgres` in the command and then enter my system password (one used at the time of logging in).

I got the following terminal output while restoring the database which shows 59 errors.
https://gist.github.com/thealphadollar/e12c97f5ad551950132b35560551992a 

On a happy note; I tested the database for the data (wow, as if I could test it for something else :walking_man: ) and data seemed to be fine and correct :)

If you can give me sometime (tests this week), and promise to have a coffee meetup with me, I'll write detailed instructions on how to import database.

**Review by thealphadollar — APPROVED** [review]
Works fine for me with the updated command I provided. :hugs: :popcorn: LGTM!
